= Recommendations to package and containerize bioinformatics software
:bibliography-database: manuscript.bibtex
:bibliography-style: apa

Authors: Yasset Perez-Riverol (1), Felipe da Veiga Leprevost (2), Pablo Moreno (1), Hervé Ménager (3), Olivier Sallou (4), Dan Søndergaard (5), Michael R. Crusoe (6), Bjorn Grunning (7), Rafael C Jimenez (8)

Affiliation:

(1) EMBL-European Bioinformatics Institute (EMBL-EBI), Hinxton, Cambridge, UK.

(2) Department of Pathology, University of Michigan, Ann Arbor, Michigan, USA.

(3) Center of Bioinformatics, Biostatistics and Integrative Biology Institut Pasteur Paris, France.

(4) Institut de Recherche en Informatique et Systèmes Aléatoires (IRISA/INRIA) - GenOuest platform, Université de Rennes 1, Rennes, France.

(5) Aarhus University, Bioinformatics Research Centre, C.F. Møllers Allé 8, Aarhus DK-8000, Denmark.

(6) Microbiology and Molecular Genetics, Michigan State University, East Lansing, MI, USA.

(7) Bioinformatics Group, Department of Computer Science, University of Freiburg, 79110, Freiburg, Germany.

(8) ELIXIR Hub, Cambridge, CB10 1SD, UK.

== Introduction

The ability to reproduce the original results of a scientific discovery has been one of the mayor challenges
in modern science. Evidences from multiple authors suggest that reproducibility in biomedical research is lower than 85% cite:[pmid24411643], while 90% of researchers agree in a reproducibility crisis cite:[pmid27225100]. One of the major drawbacks is to be able to reproduce the bioinformatics analysis, including the data processing and statistical downstream analysis cite:[pmid24204232]. Different authors has pointed three main premises for reproducible bioinformatics software deployment: (i) the use of exact versions of software and tools, (ii) open source of the code and all custom software, (iii) adopt a licence and comply with the licence of third-party dependencies cite:[pmid28751965]+[pmid24204232]. Most of these works has been focussed in the openness and availability of the tools, software, scripts and data to perform the bioinformatics analysis cite:[pmid24204232]+[pmid27415786].

However, even if source code and data are published in a public repository (e.g Github) alongside the paper as open source artifacts, they come with many dependencies, configurations, versions that make their use hard to achieve cite:[Boettiger:2015:IDR:2723872.2723882]. The build, installation and deployment of the bioinformatics solution (e.g. software or workflow) often requires internal knowledge that is missing from the published manuscript. In addition, most of the bioinformatics software is developed by different teams but used in combination with workflows, scripts or pipelines. This adds an additional layer of complexity introducing challenges such as compatibility and management of dependencies; running serial and parallel processes; and working with a broad variety of software data types and user-defined parameters.

Software containers has emerged as a powerful technology to document, distribute, and deploy bioinformatics software and workflows cite:[pmid28379341]. Containers are lightweight software components and libraries easily packaged designed to run anywhere cite:[pmid28379341], thus useful to leverage bioinformatics software reproducibility. Conda packages, Docker and Singularity are containers technologies applicable to the field of bioinformatics. By May 2018, the BioContainers cite:[pmid28379341] and BioConda cite:[207092] communities have released more than 4000 public containers facilitating the development of complex and reproducible workflows and pipelines cite:[pmid28559010]+[pmid27137889].

This manuscript describes a core set of recommendations and guidelines to improve the quality and sustainability of research software based on software packages and containers. It provides easy-to-implement recommendations that encourage adoption of packaging (e.g. conda) or containers (e.g. docker, singularity) technologies in bioinformatics and software development for research. It provides recommendations about making research software and its source code more reproducible, deployable, reusable, transparent and more compatible with other tools and software.

In this manuscript, software is broadly defined to include command line tools, graphical user interfaces, application program interfaces (APIs), infrastructure scripts and software packages (e.g. R packages).

== 1. One tool, one container.

Microservice and modular architectures cite:[7436659] is a way of breaking large software projects into smaller, independent, and loosely coupled modules. This software applications can be seen as a suite of independently deployable, small, modular components in which each tool runs a *unique* process and communicates through a well-defined, lightweight mechanism to serve a business goal cite:[7436659]. Each of these independent modules are referred to as a _container_. A container is essentially an encapsulated and immutable version of an application, coupled with the bare-minimum operating system components (e.g. dependencies) required for execution cite:[pmid28379341].

Containers should be defined as most granular as possible with the premise _one Tool, one Container_. Each container should encapsulate only one piece of software, tool that perform a unique task with a well-defined goal (e.g. sequence aligner, mass spectra identification). In order to design the level of _granularity_, different metrics should be considered:

* _Simplicity_: the encapsulated software should not be a complex environment of dependencies, tools and scripts.
* _Maintainability_: the more software is included to the container the harder it is to maintain specially when the software comes from difference sources.
* _Sustainability_: the developers of the software should be engaged or made aware to support the sustainability of the container.
* _Reusability_: a tool container should be safe to reuse by any other workflow component or task through its access interface.
* _Interoperability_: different tools should be easy to connect and exchange information
* _User’s acceptability_: tool container should encapsulate domain business process units, so it can be easier to check and use.
* _Size_: Containers should be as small as possible as big as necessary. Smaller containers are much quicker to download and therefore they can be distributed to different machines much quicker.

== 2. A package first

A software package is self-contained software including all the dependencies libraries and packages necesary to excecute the software. Some of the most popular and well-known packaging distribution system are homebrew (https://brew.sh/[https://brew.sh/]) or conda (https://conda.io/)[https://conda.io/)]. _Conda_, the most popular package manager in research software, quickly installs, runs and updates packages and their dependencies. It handles dependencies for many languages such as C, C++, R, Java and of course Python tools. In addition, _Conda_ has joined to other popular packages manager systems such as Gentoo, BSD Ports, MacPorts, and Homebrew which build packages from source instead of installing from a pre-built binary. The field of computational biology has built an strong community (https://bioconda.github.io/)[https://bioconda.github.io/)].

Bioconda is a channel for the conda package manager specialised in bioinformatics software. You can create a _Conda_ package by defining a _BioConda_ recipe (**Box 1**). This recipe (https://github.com/bioconda/bioconda-recipes[https://github.com/bioconda/bioconda-recipes]) includes enough information about the dependencies, the LICENSE and fundamental metadata to find, retrieve and use the package (see *Recomendation X*). Each package added to Bioconda is also made available as a Docker container via Quay.io and displayed in the BioContainers registry cite:[pmid28379341]. The BioContainers registry displays Docker containers as well as containers from other technologies such as rkt and Singularity cite:[pmid28494014] making easier to find a discover packages and related containers.

**************************

package:
  name: abawaca
  version:1.00

source:
  fn: abawaca-1.00.tar.gz
  url: https://github.com/CK7/abawaca/archive/v{{ version }}.tar.gz
  sha256: 57465bb291c3a9af93605ffb11d704324079036205e5ac279601c9e98c467529

build:
  number: 0

requirements:
  build:
    - gcc   # [not osx]
    - llvm  # [osx]

  run:
    - libgcc    # [not osx]

test:
  commands:
    - abawaca 2>&1 | grep Usage

about:
  home: 'https://github.com/CK7/abawaca'
  license: "open source"
  summary: "abawaca is a binning program for metagenomics"

extra:
  maintainers:
    - keuv-grvl
  identifiers:
    - doi:10.1038/nature14486
**************************

Box 1: BioConda Recipe for "abawaca", a binning program for metagenomics.

== 3. Versions should be explicit, and consider both the tool version and the container version

The tool or software wrapped inside the container should be fixed explicitly to a defined version through the mechanism available by the package manager or install method used (**Box 2**). The version used for this main software should be included in both, the metadata of the container (for findability reasons) and the container tag. The tag and metadata of the container should also include a versioning number for the container itself, meaning that the tag could look like `&lt;version-of-the-tool&gt;_cv&lt;version-of-the-container&gt;`. The container version, which does not track the tool changes but the container, should be versioned through semantic versioning to signal its backward compatibility.

**************************
FROM biocontainers/biocontainers:latest

LABEL base_image="biocontainers:latest"

LABEL version="3"

LABEL software="Comet"

LABEL software.version="2016012"

LABEL about.summary="an open source tandem mass spectrometry sequence database search tool"

LABEL about.home="http://comet-ms.sourceforge.net/"

LABEL about.documentation="http://comet-ms.sourceforge.net/parameters/parameters_2016010/"

LABEL about.license_file="http://comet-ms.sourceforge.net/"

LABEL about.license="SPDX:Apache-2.0"

LABEL extra.identifiers.biotools="comet"

LABEL about.tags="Proteomics"

################## MAINTAINER ######################

MAINTAINER Felipe da Veiga Leprevost <felipe@leprevost.com.br>

################## INSTALLATION ######################

USER biodocker

RUN ZIP=comet_binaries_2016012.zip && \
  wget https://github.com/BioDocker/software-archive/releases/download/Comet/$ZIP -O /tmp/$ZIP && \
  unzip /tmp/$ZIP -d /home/biodocker/bin/Comet/ && \
  chmod -R 755 /home/biodocker/bin/Comet/* && \
  rm /tmp/$ZIP

RUN mv /home/biodocker/bin/Comet/comet_binaries_2016012/comet.2016012.linux.exe /home/biodocker/bin/Comet/comet

ENV PATH /home/biodocker/bin/Comet:$PATH

WORKDIR /data/
**************************

Box 2: BioContainers recipe for comet software. The metadata container the license of the software.

If a copy is done via a git clone or equivalent, a specific tagged version should be copied, never the _latest_ branch. Cloning the latest branch will copy the latest code making difficult to reproduce an operation since the latest branch might suffer constant changes. Upstream authors should be asked to create a release if not available. In the worst case, the HEAD commit id of the clone should be used as the tool version for the container.

== 4. Eschew ENTRYPOINT

Is a well-known feature of Docker that the entry-point of the container can be over-write by definition (e.g, ENTRYPOINT ["/bin/ping"]). The **ENTRYPOINT** specifies a command that will always be executed when the container starts. Even when the ENTRYPOINT helps the user get _default_ behaviour for the tool, it is not well-recommended for reproducibility efforts of the original results or workflow. By specifinying the name of the tool in the container (not ENTRYPOINT) the user (e.g. workflow) recognize and trace which tool is used within the container.

== 5. Relevant tools and software should be executable and in the path

If for some reason the container needs to expose more than a single executable or script (for instance, EMBOSS or other packages with many executables), these should always be executable and be available in the container's default path. This will be mostly always the case by default for everything that is installed via a package manager (apt-get, pip, R install.package, etc.) and you won't need to worry if that is the case, but if you are adding tailored made scripts or others that are not installed through a manager, then you need to take care of this. This will facilitate the use of the container as an environment or to specify alternative commands to the main entrypoint easily.

== 6. Reduce the size of your container as much as possible

Since containers are being constantly being pushed and pulled over the internet, their size matters. There are many tips to reduce the size of your container in build time:
- Avoid installing "recommended" packages in apt based systems.
- Do not keep build tools in the image: this includes compilers and development libraries that will seldomly, if not at all, used in runtime when your container is being used by others. For instance, packages like gcc can use several hundred megabytes. This also applies to tools like git, wget or curl, which you might have used to retrieve software during container buildtime, but are not needed for runtime.
- Make sure you clean caches, unneeded downloads and temporary files.
- In Dockerfiles, combine multiple RUNs so that the initial packages installations and the final deletions (of compilers, development libraries and caches/temporary files) are left within the same layer.
- If installing or cloning from a git repo, use shallow clones, which for large repos will save a lot of space.

== 7. Choose a base image wisely.

One of the decisions that will most likely impact on your final container image size will be your base image. If you can, start with a lightweight base image such as Alpine or similar, always at a fixed version. If installing your software on top of such a minimal operating system doesn't work out well, only then move to a larger, stock-image where installation of your tool software might be simpler (such as Ubuntu). Preferring stock images means that many other people will be using them and that your container will be pulled faster as shared layers are more likely. Always aim to have predefined images from where you choose (always the same Alpine version as first choice and always the same Ubuntu version as second choice), so that most of your containers share that base image.

== 8. Add some testing logic

If others want to build locally your container, want to rebuild it later on with an updated base image, want to integrate it to a continuous integration system for building it or for many other reasons, users might want to test that the built container still serves the function for which it was originally designed. For this is useful to add to the container some testing logic inside it (in the form of a bash script for instance) in a standard location (here we propose a file called `runTest.sh`, executable and in the path) which includes all the logic for:
- Installing any packages that might be needed for testing, such as wget for instance to retrieve example files for the run.
- Obtain sample files for testing, which might be for instance an example data set from a reference archive.
- Run the software that the container wraps with that data to produce and output inside the container.
- Compare the produced output and exit with an error code if the comparison is not successful.
This file containing the testing logic is not meant to be executed during container buildtime, so the retrieved data/packages don't increase the size of the container when this is executed once the container is built. This means that, because the file is inside the container, any user who has built the container or downloaded the container image can check that the container is working adequately by executing `runTest.sh` through the container.

== 9. Check the license of the software

When adding software or data in a container, always check their license. A free to use license is not always a free to distribute or copy. License _must_ always be explicitly defined in your labels and depending on license, you must also include a copy of the license with the software. Same care must be applied to included data.
If license is not specified, you should ask upstream author to provide a license.

== 10. Make you package or container findable

Biomedical research and bioinformatics demands more efforts to make bioinformatics software and data more Findable, Accessible, Interoperable, and Reusable (FAIR Principles) cite:[pmid26978244]. Leveraging those principles, we recommend to the bioinformatics community and software developers to make their containers and package more findable. In order to make your package available we recommend the following steps:

- Annotate your package or container with some metadata that allows users (e.g. biologiests, bioinformatians) to find them.

- Make your package/container available. We recommend that as a developer you should make the recipe of how to build your containers available for others, including i) the source code or binaries of the original tools; ii) configuration settings and test data.

- Register your package/container with some of the existing bioinformatics intitiatives such as BioContainers cite:[pmid28379341], bio.tools cite:[pmid26538599] or BioConda cite:[207092]. Al these resources are fully connected and exchange metadata and information using different APIs and a common identifier system.

== Conclusions

If you are involved in scientific research and have not containers or software packages  before, we recommend that you explore its potential as soon as possible cite:[pmid28379341]. As with many tools, a learning curve lays ahead, but several basic yet powerful features are accessible even to the beginner and may be applied to many different use-cases. To conclude, we would like to recommend some examples of bioinformatics containers in Biocontainers (Table 1) and some useful training materials, including workshops, online courses, and manuscripts (Table 2).

== References

bibliography::[]
